src/jmteb/evaluators/clustering/evaluator.py

```diff
@@ -22,80 +23,91 @@
+            "BisectingKMeans": lambda: BisectingKMeans(n_clusters=n_clusters),
+            "Birch": lambda: Birch(n_clusters=n_clusters),
+        }
 
         logger.info("Fitting clustering model...")
-        clustering_models = []
-        if self.k_means_clustering:
-            clustering_models.append(
-                MiniBatchKMeans(
-                    n_clusters=n_clusters,
-                    n_init="auto",
-                )
-            )
-        if self.hierarchical_clustering:
-            clustering_models.extend(
-                [
-                    AgglomerativeClustering(n_clusters=n_clusters),
-                    BisectingKMeans(n_clusters=n_clusters),
-                    Birch(n_clusters=n_clusters),
-                ]
-            )
-        for clustering_model in clustering_models:
-            _results: dict[str, float] = {}
-            clustering_model.fit(embeddings)
-            y_pred = clustering_model.labels_
+        val_results = {}
+        for model_name, model_constructor in model_constructors.items():
+            val_results[model_name] = self._evaluate_clustering_model(val_embeddings, val_labels, model_constructor())
+        optimal_clustering_model_name = sorted(
+            val_results.items(),
+            key=lambda res: res[1][self.main_metric],
+            reverse=True,
+        )[0][0]
 
-            # compute metric
-            h_score, c_score, v_score = homogeneity_completeness_v_measure(
-                labels_pred=y_pred, labels_true=np.array(labels)
-            )
-            clustering_model_name = type(clustering_model).__name__
-            _results.update(
-                {
-                    "v_measure_score": v_score,
-                    "homogeneity_score": h_score,
-                    "completeness_score": c_score,
-                }
+        test_results = {
+            optimal_clustering_model_name: self._evaluate_clustering_model(
+                test_embeddings,
+                test_labels,
+                model_constructors[optimal_clustering_model_name](),
             )
-            results[clustering_model_name] = _results
+        }
 
         return EvaluationResults(
             metric_name=self.main_metric,
-            metric_value=max([v[self.main_metric] for v in results.values()]),
-            details=results,
+            metric_value=test_results[optimal_clustering_model_name][self.main_metric],
+            details={
+                "optimal_clustering_model_name": optimal_clustering_model_name,
+                "val_scores": val_results,
+                "test_scores": test_results,
+            },
+        )
+
+    @staticmethod
+    def _evaluate_clustering_model(
+        embeddings: np.ndarray, y_true: list[int], clustering_model: ClusterMixin
+    ) -> dict[str, float]:
+        clustering_model.fit(embeddings)
+        y_pred = clustering_model.labels_
```

Comment:
`ClusterMixin` クラス自体にはは `fit` や `labels_` が定義されていないので IDE で warning が出てしまいました。
（Mixin クラスなのでしょうがないのですが）
下のようにすると `fit_predict` が `ClusterMixin` に定義されているので、ちょっとスッキリします。

```suggestion
        y_pred = clustering_model.fit_predict(embeddings)
```